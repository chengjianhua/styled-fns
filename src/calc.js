import invariant from 'invariant'

const CHAR_CODE_DOT = 44

const CHAR_CODE_ZERO = 48
const CHAR_CODE_NINE = 57

const splitValueUnit = string => {
  for (let i = 0, length = string.length; i < length; i++) {
    const charCode = string.charCodeAt(i)

    if (
      charCode === CHAR_CODE_DOT ||
      (charCode >= CHAR_CODE_ZERO && charCode <= CHAR_CODE_NINE)
    )
      // eslint-disable-next-line no-continue
      continue

    return {unit: string.slice(i, length), value: string.slice(0, i)}
  }

  return null
}

const calcForTemplate = (strings, variables) => {
  let unit = null

  const expression = strings.reduce((acc, string) => {
    const variable = variables.shift()

    if (variable) {
      const unitValue = splitValueUnit(variable)

      if (!unitValue) return acc + string + variable

      if (unit === null) {
        unit = unitValue.unit
      }

      // console.log({
      //   variable,
      //   value: unitValue.value,
      //   unit,
      //   _unit: unitValue.unit,
      // });

      invariant(
        unitValue.unit === unit,
        `the calculation expression should contain only one sort of unit`,
      )

      return acc + string + unitValue.value
    }

    return acc + string
  }, '')

  // console.log('expression: ', expression);

  // eslint-disable-next-line no-new-func
  const calculate = new Function(`return ${expression}`)

  return calculate() + unit
}

// const UNIT_REG = new RegExp('(?<=\\d)[^\\d.()\\+\\-\\*\\/\\s]+(?=\\s)+', 'g');
// const calcForString = string => {
//   const unit = UNIT_REG.exec(string);
//   console.log({ string, unit });
// };

export default function calc(strings, ...variables) {
  // NOTE: don't support the full string expression without variables right now

  // if (variables.length === 0) {
  //   return calcForString(strings[0]);
  // }

  // if (strings.length === 1 || typeof strings === 'string') {
  //   return calcForString(strings);
  // }

  invariant(
    !(
      variables.length === 0 ||
      (strings.length === 1 || typeof strings === 'string')
    ),
    'Do not support the full string expression without variables right now temporarily',
  )

  try {
    return calcForTemplate(strings, variables)
  } catch (error) {
    const {message} = error

    if (error instanceof SyntaxError) {
      error.message = `You may not wrap your variable has unit with \`\${}\`\n\n${message}`
    }

    throw error
  }
}
