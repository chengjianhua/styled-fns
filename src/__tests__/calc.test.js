import calc from '../calc'

describe('calc', () => {
  it('should return right result for template with variables', () => {
    expect(calc`(${'100px'} * 100) + ${'1px'}`).toBe('10001px')
    expect(calc`(${'100px'} * 100) / 1`).toBe('10000px')
  })

  it('should throw exception if occurs more than two sort of units in expression', () => {
    expect(() => {
      // eslint-disable-next-line
      calc`(${'100px'} * 100) + ${'1rem'}`
    }).toThrowErrorMatchingSnapshot()
  })

  it('should return right result for variables without unit, and return value is type of `Number`', () => {
    expect(calc`${'100'} * 100`).toBe(10000)
  })

  it('should throw exception if input is full string without variables', () => {
    expect(() => {
      // eslint-disable-next-line
      calc`(1.5px * 100) + 100px`
    }).toThrowErrorMatchingSnapshot()
  })

  it('should throw custom error message while input contains variables but variable has unit without being wrapped', () => {
    expect(() => {
      // eslint-disable-next-line
      calc`(${'1.5px'} * 100) + 100px`
    }).toThrowErrorMatchingSnapshot()
  })
})
