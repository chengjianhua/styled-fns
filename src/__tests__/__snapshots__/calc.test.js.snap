// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calc should throw custom error message while input contains variables but variable has unit without being wrapped 1`] = `
"You may not wrap your variable has unit with \`\${}\`

Invalid or unexpected token"
`;

exports[`calc should throw exception if input is full string without variables 1`] = `"Do not support the full string expression without variables right now temporarily"`;

exports[`calc should throw exception if occurs more than two sort of units in expression 1`] = `"the calculation expression should contain only one sort of unit"`;
